Add opt-in non-blocking group-by combine operator

xiangfu0 opened the pull request:
## Summary

Adds an opt-in non-blocking server-side GROUP BY combine operator. Each worker accumulates into an exclusively owned
`SimpleIndexedTable`, then the workers combine their tables through an atomic tournament that merges the smaller table
into the larger one.

The existing `CONCURRENT` implementation remains the server default and rollout gate. The non-blocking path is used
only for general GROUP BY queries with `ORDER BY` after an operator enables it in server configuration. Sorted
safe-trim operators still take precedence, and GROUP BY queries without `ORDER BY` use the shared concurrent table.

The experimental partitioned implementation was removed after review found correctness and peak-memory risks that
were not justified by its additional complexity.

## Motivation

`GroupByCombineOperator` sends every worker's upserts through one shared `ConcurrentIndexedTable`. High-cardinality
queries spend substantial time contending on that table. The non-blocking implementation moves the hot accumulation
loop to ordinary per-worker `HashMap`s and performs a bounded reduction after segment processing.

## Design

- Build one thread-confined `SimpleIndexedTable` per active combine worker.
- Exchange completed tables through an `AtomicReference` without a central merge lock.
- Merge smaller tables into larger tables to reduce repeated hash-map growth.
- Transfer existing `Key` and `Record` ownership while destructively draining source maps, reducing allocation and
  peak retained memory during reduction.
- Stop partially drained tournament merges after a worker failure and discard both exclusively owned tables.
- Avoid capturing map-update lambdas in the simple-table hot path.
- Reuse `TableResizer` candidate storage instead of allocating one wrapper per examined record.
- Preserve cancellation checks and aggregate trim/resize statistics across table merges.
- Preserve the legacy unordered-table insertion behavior used by downstream reducers.
- Keep `CONCURRENT` as the default because per-worker tables can require more peak memory.
- Prevent a query option from bypassing the server rollout gate.
- Route no-`ORDER BY` GROUP BY queries to `CONCURRENT`; worker-local LIMIT admission can otherwise discard partial
  aggregates before the global merge.

## Rollout and usage

Enable the algorithm on a server after validating representative query cardinality, execution-thread count, and
memory headroom:

```properties
pinot.server.query.executor.group.by.algorithm=NON-BLOCKING
```

With that server gate enabled, a query can explicitly select the optimized path:

```sql
SET groupByAlgorithm = 'NON-BLOCKING';
SELECT d1, d2, SUM(m1), MAX(m2)
FROM testTable
GROUP BY d1, d2
ORDER BY SUM(m1) DESC
LIMIT 500;
```

A query can fall back to the shared implementation without changing server configuration:

```sql
SET groupByAlgorithm = 'CONCURRENT';
SELECT d1, d2, SUM(m1), MAX(m2)
FROM testTable
GROUP BY d1, d2
ORDER BY SUM(m1) DESC
LIMIT 500;
```

If the server remains configured as `CONCURRENT`, a query cannot opt itself into `NON-BLOCKING`.

## Performance

JMH on JDK 25, 8 combine threads, 4 forks, and 3 measured iterations per fork. The harness explicitly rejects unknown
algorithm names and was invoked with `-p '_algorithm=CONCURRENT,NON-BLOCKING'`. The benchmark query is:

```sql
SELECT d1, d2, SUM(m1), MAX(m2)
FROM testTable
GROUP BY d1, d2
ORDER BY SUM(m1) DESC
LIMIT 500
```

| Workload | CONCURRENT | NON-BLOCKING | Speedup | Allocation reduction |
|---|---:|---:|---:|---:|
| 100 segments x 100K records, repeated 10-dataset pool | 4,124.7 ms/op | 1,689.6 ms/op | **2.44x** | **20.8%** |
| 10 segments x 1M records, 10 distinct datasets | 6,153.2 ms/op | 1,562.8 ms/op | **3.94x** | **22.5%** |
| 100 segments x 100K records, 100 distinct datasets (~6.32M groups) | 6,184.0 ms/op | 1,386.7 ms/op | **4.46x** | **26.4%** |

Normalized allocation:

| Workload | CONCURRENT | NON-BLOCKING |
|---|---:|---:|
| 100 x 100K, 10-dataset pool | 2,715,889,806 B/op | 2,151,570,302 B/op |
| 10 x 1M, 10 datasets | 2,886,638,027 B/op | 2,236,964,905 B/op |
| 100 x 100K, 100 datasets | 2,925,703,405 B/op | 2,154,772,481 B/op |

## Validation

Targeted correctness coverage includes legacy and non-blocking inter-segment queries, trim behavior, cancellation,
rollout/default selection, a forced four-worker CAS tournament, deterministic partial-merge failure, no-`ORDER BY`
partial aggregates, and the gapfill regressions found by the JDK 25 CI shard.

```bash
TZ=UTC ./mvnw -pl pinot-core \
  -Dtest=GapfillQueriesTest,GapfillQueriesScalabilityTest,NonblockingGroupByCombineOperatorTest,CombinePlanNodeTest,IndexedTableTest,DeterministicIndexedTableTest,TableResizerTest,SortedGroupByCombineOperatorsTest,InstancePlanMakerImplV2Test,InterSegmentGroupBySingleValueQueriesTest,InterSegmentNonBlockingGroupBySingleValueQueriesTest \
  test
```

Result on Temurin 25: **247 tests passed**.

Required module checks also pass:

```bash
./mvnw spotless:apply -pl pinot-common,pinot-core,pinot-spi,pinot-perf
./mvnw checkstyle:check -pl pinot-common,pinot-core,pinot-spi,pinot-perf
./mvnw license:format -pl pinot-common,pinot-core,pinot-spi,pinot-perf
./mvnw license:check -pl pinot-common,pinot-core,pinot-spi,pinot-perf
```



## CI

All **11/11** required GitHub Actions checks passed on exact head `bb9a6b212a2f63b607272a69a78c8f7ade5be7ad`, including both unit shards, both integration shards, quickstart, linter, binary compatibility, and all four compatibility regressions.

ankitsultana commented:
@xiangfu0 : are you folks also planning to run some benchmarks? And any other ideas you are already trying out?

wirybeaver commented:
@xiangfu0 Glad to see the master is working on the partitioned groupBy algorithm. Do we plan to support disk spilled mode? https://github.com/apache/pinot/issues/12080

xiangfu0 commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/server/ServerPlanRequestUtils.java:
```diff
@@ -92,7 +92,7 @@ public static OpChain compileLeafStage(OpChainExecutionContext executionContext,
    *
    * @param executionContext the execution context used by the leaf-stage execution engine.
-   * @param stagePlan the distribute stage plan on the leaf.
+   * @param stagePlan        the distribute stage plan on the leaf.
```
revert changes on this PR

xiangfu0 committed:
Add non-blocking and partitioned group-by combine operators

Introduce two new server-level group-by combine algorithms selectable
via the `groupByAlgorithm` query option:

- NONBLOCKING: Uses a lock-free ConcurrentHashMap to aggregate results
  across segments without holding a global lock, reducing contention
  under high concurrency.

- PARTITIONED: Hash-partitions group keys across independent tables so
  that each partition can be aggregated and trimmed in parallel with
  zero cross-partition synchronization. Includes per-partition
  thread-local accumulation, cached key hashing, and parallel
  finish/trim to minimize overhead for high-cardinality queries.

Also adds:
- `normalizeGroupByAlgorithm` helper in QueryOptionsUtils
- `IndexedTable.transferFrom` for bulk partition merging
- JMH benchmark for comparing combine operator throughput
- Unit and integration tests for both new algorithms

xiangfu0 commented on .github/workflows/scripts/.pinot_quickstart.sh:
```diff
@@ -80,7 +80,7 @@ echo "Building Pinot to JAVA 11 source code Using JDK ${JAVA_VER}"
 for i in $(seq 1 2)
 do
-  mvn clean install -B -ntp -T1C -DskipTests -Pbin-dist -Dmaven.javadoc.skip=true -Djdk.version=11
+  mvn clean install -B -ntp -T1C -DskipTests -Pbin-dist,sqlparser -Dmaven.javadoc.skip=true -Djdk.version=11
```
Reverted — this file is no longer modified in the PR.

xiangfu0 committed:
Rename DEFAULT algorithm to CONCURRENT, use query option constant in tests

- GroupByCombineOperator.ALGORITHM: "DEFAULT" → "CONCURRENT" (descriptive name
  for the shared ConcurrentHashMap-based operator; "DEFAULT" was misleading since
  it is no longer the default)
- CombinePlanNode: update comment to reference CONCURRENT
- InterSegmentNonBlockingGroupBySingleValueQueriesTest: use
  QueryOptionKey.GROUP_BY_ALGORITHM constant instead of raw string, use
  NonblockingGroupByCombineOperator.ALGORITHM / GroupByCombineOperator.ALGORITHM
  constants instead of hardcoded strings, add tests for the CONCURRENT fallback path

xiangfu0 committed:
Fix review findings: instance-level config, encapsulation, Javadoc

- Add Server.GROUP_BY_ALGORITHM config key so operators can set a fleet-wide
  default algorithm without modifying individual queries (MAJOR: no server-level
  rollback path). InstancePlanMakerImplV2 reads it at init and falls back to
  NON-BLOCKING. Per-query SET groupByAlgorithm=... still overrides.

- Revert GroupByCombineOperator fields back to private (only _indexedTable was
  needed by the subclass). Replace direct field access with three protected
  exchange methods: stealSharedTable(), depositSharedTable(), hasSharedTable().
  NonblockingGroupByCombineOperator now uses these methods instead of reaching
  into the parent's internal state (MINOR: overly wide protected field scope).

- Make EXPLAIN_NAME protected in GroupByCombineOperator; remove the duplicate
  constant and the redundant toExplainString() override from the subclass
  (MINOR: duplicate constant).

- Add Javadoc to IndexedTable.iterator() documenting the pre-finish semantics
  and to absorbTrimStats() documenting the non-thread-safe ownership requirement
  (MINOR: undocumented contract change).

xiangfu0 committed:
Fix PartitionedGroupByCombineOperator: Knuth hash, trim loop, and comment accuracy

- Replace h^h>>>16 secondary hash with Knuth multiplicative hash (0x9E3779B9) in
  partitionFor() to eliminate HashMap bucket clustering in partition-local tables
- Fix trim storm in maybeTrimLocalPartitions: loop until actualTotal drops below
  threshold instead of trimming only the largest partition once
- Clarify _indexedTable comment to accurately reflect volatile DCL access pattern

xiangfu0 committed:
Optimize non-blocking group-by combine

xiangfu0 committed:
Make group-by contention test deterministic

xiangfu0 committed:
Preserve non-blocking group-by compatibility
